src/sp/it/pl/util/async/future/Fut.kt

```diff
@@ -0,0 +1,150 @@
+package sp.it.pl.util.async.future
+
+import javafx.scene.control.ProgressIndicator
+import javafx.util.Duration
+import mu.KLogging
+import sp.it.pl.main.APP
+import sp.it.pl.util.async.FX
+import sp.it.pl.util.async.NEW
+import sp.it.pl.util.functional.kt
+import sp.it.pl.util.async.sleep
+import sp.it.pl.util.functional.Try
+import sp.it.pl.util.functional.invoke
+import sp.it.pl.util.functional.toUnit
+import java.util.Optional
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.ExecutionException
+import java.util.concurrent.Executor
+import java.util.function.Consumer
+
+/**
+ * Future monad implementation.
+ *
+ * Oriented for practicality * ease of use, not specification (monadic laws) or robustness (API completeness).
+ */
+class Fut<T>(private var f: CompletableFuture<T>) {
+
+    fun cancel(): Unit = f.cancel(true).toUnit()
+
+    @JvmOverloads
+    fun <R> then(executor: Executor = defaultExecutor, block: (T) -> R) = Fut<R>(f.thenApplyAsync(block.logging(), executor.kt))
+
+    @JvmOverloads
+    fun <R> then(executor: Executor = defaultExecutor, block: Fut<R>): Fut<R> = Fut(f.thenComposeAsync( { block.f }, executor.kt))
+
+    /** [then] which sleeps the specified duration on [NEW]. */
+    fun <R> thenWait(time: Duration) = then(NEW) { sleep(time) }
+
+    /** [then] with [FX] executor. Intended for simple and declarative use of asynchronous computation from ui. */
+    infix fun <R> ui(block: (T) -> R) = then(FX, block)
+
+    /** [then] which returns the original value. Intended for side-effects. */
+    fun use(executor: Executor = defaultExecutor, block: (T) -> Unit) = then(executor) { block(it); it }
+
+    /** Legacy version of [use] for Java taking a [Consumer]. */
+    @JvmOverloads
+    fun useBy(executor: Executor = defaultExecutor, block: Consumer<in T>) = then(executor) { block(it); it }
+
+    /**
+     * Returns new future, which sets progress to 0 on fx thread, then executes
+     * this future and then sets progress to 1, again on fx thread.
+     *
+     * Note that when chaining futures, the position within chain decides when
+     * does the progress reach 1. It will not be at the end of the chain, but
+     * at the position of this method in it. The progress is set to 0 always at
```

Comment:
I don't see through this

Reply:
Ugh, I will provide proper doc for the entire class

Reply:
I improved documentation, but ultimately progress handling is a larger issue and blocked. Out of scope.